### PR TITLE
Integrate Grouping Service

### PR DESCRIPTION
**Pull request by NereusWB922**
### Summary:

Fixes part of #306 

#### Type of change:

- ✨ New Feature/ Enhancement
-
### Changes Made:

- Refactor `IssueDataTable` to use `GroupingContextService` for data grouping
- Refactor `IssueViewerComponent` to use `GroupingContextService` for grouping
- Refactor `CardViewComponent` to render dynamic header for different `Group` type
- Refactor `HiddenUsersComponent` to render different `Group` type
- Implement group by selection form in `FilterBarComponent`

### Proposed Commit Message:

```
Integrate Grouping Service

Implement GroupBy feature to allow users to group the issues/prs
based on different criteria such as milestone, status and etc.

Let's integrate the grouping service in the components.
```

### CODE REVIEW

File: src/app/issues-viewer/issues-viewer.component.ts [hunk 1]
```diff
@@ -96,25 +104,27 @@ export class IssuesViewerComponent implements OnInit, AfterViewInit, OnDestroy {
...
+
+    if (!isGroupPresent) {
+      this.hiddenGroups.push(target);
     }
   }
 
-  private removeHiddenUser(assignee: GithubUser) {
-    this.hiddenAssignees = this.hiddenAssignees.filter((user) => user !== assignee);
+  private removeHiddenGroup(target: Group) {
+    this.hiddenGroups = this.hiddenGroups.filter((group) => !group.equals(target));
```
Comment (Arif-Khalid) [inline src/app/issues-viewer/issues-viewer.component.ts:128]: The naming of these functions could be made clearer. As they are they imply updating the group and removing the group but is unclear what the update is or what the removal is from.
A clearer name would be something like ``addToHiddenGroups`` and ``removeFromHiddenGroups``

File: src/app/issues-viewer/issues-viewer.component.ts [hunk 2]
```diff
@@ -40,13 +43,18 @@ export class IssuesViewerComponent implements OnInit, AfterViewInit, OnDestroy {
...
+    public milestoneService: MilestoneService,
+    public groupingContextService: GroupingContextService
   ) {
     this.repoChangeSubscription = this.viewService.repoChanged$.subscribe((newRepo) => {
       this.issueService.reset(false);
       this.labelService.reset();
       this.initialize();
     });
+
+    this.groupByChangeSubsription = this.groupingContextService.currGroupBy$.subscribe((newGroupBy) => {
```
Comment (MadLamprey) [inline src/app/issues-viewer/issues-viewer.component.ts:55]: Minor typo in `groupByChangeSubscription`

**Review by nknguyenhc — DISMISSED** [review]
LGTM!

**Review by Arif-Khalid — COMMENTED** [review]
An excellent PR which lays the groundwork for switching by group. Perhaps the currGroupBy variable could be stored and updated as a filter since I would consider it a filter variable. This would reduce duplicated code for enhancements regarding filters such as storing filters in the url.

Also a minor complaint about naming, but otherwise a great implementation which no doubt took a lot of work! 😀

**Review by MadLamprey — COMMENTED** [review]
LGTM!

**Review by MadLamprey — APPROVED** [review]
LGTM!

**Review by Arif-Khalid — APPROVED** [review]
LGTM!

**Pull request merged**